<launch>

  <!-- Name of the node. All topics will be advertised under this name. -->
  <arg name="base_name"          default="pico_flexx"/>
  <!-- ID of the sensor that should be used. IDs of all connected devices are listed on startup. -->
  <arg name="sensor"             default=""/>
  <!-- ID of the use case. A list of supported use cases is listed on startup. -->
  <arg name="use_case"           default="1"/>
  <!-- Enable or disable automatic expusure. -->
  <arg name="automatic_exposure" default="true"/>
  <!-- Exposure time. Only for manual exposure. -->
  <arg name="exposure_time"      default="1000"/>
  <!-- Exposure mode (0 = manual, 1 = automatic) -->
  <arg name="exposure_mode"      default="1"/>
  <!-- Stream 2 Exposure time. Only for manual exposure. -->
  <arg name="exposure_time_stream2" default="1000"/>
  <!-- Stream 2 Exposure mode (0 = manual, 1 = automatic) -->
  <arg name="exposure_mode_stream2"  default="1"/>
  <!-- Maximum allowed noise. Data with higher noise will be filtered out. -->
  <arg name="max_noise"          default="0.07"/>
  <!-- Filter level (0 = Off, 200 = Legacy, 255 = Full) -->
  <arg name="filter_level"       default="200"/>
  <!-- Range of the 16-Bit mono image which should be mapped to the 0-255 range of the 8-Bit mono image. The resulting range is `range_factor` times the standard deviation arround mean. -->
  <arg name="range_factor"       default="2.0"/>
  <!-- Queue size for publisher. -->
  <arg name="queue_size"         default="2"/>
  <!-- Publish a static TF transform for the optical frame of the camera. -->
  <arg name="publish_tf"         default="false" />
  <!-- Base name of the tf frames. -->
  <arg name="base_name_tf"       default="$(arg base_name)"/>
  <!-- Machine on with the nodes should run. -->
  <arg name="machine"            default="localhost"/>
  <!-- Whether the machine for localhost should be defined our not. Disable this if the launch file is included somewhere where machines are already defined. -->
  <arg name="define_machine"     default="true"/>
  <!-- Name of the nodelet manager. -->
  <arg name="nodelet_manager"    default="$(arg base_name)"/>
  <!-- Whether to start a nodelet manager our not. Disable this if a different nodelet manager should be used. -->
  <arg name="start_manager"      default="true"/>

  <!-- Default machine definition. -->
  <machine name="localhost" address="localhost" if="$(arg define_machine)"/>

  <!-- Static TF publisher. -->
  <node pkg="tf" type="static_transform_publisher" name="$(arg base_name)_static_tf" machine="$(arg machine)" if="$(arg publish_tf)"
        args="0 0 0 0 0 -1.57079632679489661923 $(arg base_name_tf)_link $(arg base_name_tf)_optical_frame 100" />

  <!-- Nodelet manager. -->
  <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager"
        if="$(arg start_manager)" machine="$(arg machine)" output="log"/>

  <!-- pico_flexx_driver nodelet. -->
  <node pkg="nodelet" type="nodelet" name="$(arg base_name)_driver" machine="$(arg machine)"
        args="load pico_flexx_driver/pico_flexx_nodelet $(arg nodelet_manager) pico_flexx/points:=/points pico_flexx/image_depth:=/image_depth pico_flexx/image_mono8:=/image_nir "
        respawn="true" output="log">
    <param name="base_name"          type="str"    value="$(arg base_name)"/>
    <param name="sensor"             type="str"    value="$(arg sensor)"/>
    <param name="use_case"           type="int"    value="$(arg use_case)"/>
    <param name="automatic_exposure" type="bool"   value="$(arg automatic_exposure)"/>
    <param name="exposure_time"      type="int"    value="$(arg exposure_time)"/>
    <param name="exposure_mode"      type="int"    value="$(arg exposure_mode)"/>
    <param name="exposure_time_stream2" type="int"    value="$(arg exposure_time_stream2)"/>
    <param name="exposure_mode_stream2" type="int"    value="$(arg exposure_mode_stream2)"/>
    <param name="max_noise"          type="double" value="$(arg max_noise)"/>
    <param name="filter_level"       type="int"    value="$(arg filter_level)"/>
    <param name="range_factor"       type="double" value="$(arg range_factor)"/>
    <param name="queue_size"         type="int"    value="$(arg queue_size)"/>
    <param name="base_name_tf"       type="str"    value="$(arg base_name_tf)"/>
  </node>

</launch>
